lib/AST/Decl.cpp

```diff
@@ -6998,6 +6998,49 @@ StringRef AbstractFunctionDecl::getInlinableBodyText(
   return extractInlinableText(getASTContext().SourceMgr, body, scratch);
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+/// A uniqued list of derivative function configurations.
+struct AbstractFunctionDecl::DerivativeFunctionConfigurationList
+    : public llvm::SetVector<AutoDiffConfig> {
+  // Necessary for `ASTContext` allocation.
+  void *operator new(
+      size_t bytes, ASTContext &ctx,
+      unsigned alignment = alignof(DerivativeFunctionConfigurationList)) {
+    return ctx.Allocate(bytes, alignment);
+  }
+};
+
+void AbstractFunctionDecl::prepareDerivativeFunctionConfigurations() {
+  if (DerivativeFunctionConfigs)
+    return;
+  auto &ctx = getASTContext();
+  DerivativeFunctionConfigs = new (ctx) DerivativeFunctionConfigurationList();
+  // Register an `ASTContext` cleanup calling the list destructor.
+  ctx.addCleanup([this]() {
+    this->DerivativeFunctionConfigs->~DerivativeFunctionConfigurationList();
+  });
+}
+
+ArrayRef<AutoDiffConfig>
```

Comment:
Note: even though `@differentiable` and `@derivative` attributes do not currently store result indices, I decided to use `AutoDiffConfig` (which contains result indices) to represent AST derivative function configurations. I also tried creating a new "ASTAutoDiffConfig" struct containing just parameter indices and derivative generic signature, but decided against the code duplication.

I think the direction is to eventually plumb result indices through the differentiation system (TF-1038), so preemptively using result indices (`AutoDiffConfig`) seems good.